packages/yew-agent/src/reactor/hooks.rs

```diff
@@ -0,0 +1,279 @@
+        *on_output_ref = on_output;
+    }
+
+    let tx = use_memo((worker_state, ctr.inner), |(state, _ctr)| {
+        let bridge = state.create_bridge();
+
+        let (tx, mut rx) = bridge.split();
+
+        spawn_local(async move {
+            while let Some(m) = rx.next().await {
+                let on_output = on_output_ref.borrow().clone();
+                on_output(ReactorEvent::<R>::Output(m));
+            }
+
+            let on_output = on_output_ref.borrow().clone();
+            on_output(ReactorEvent::<R>::Finished);
+        });
+
+        RwLock::new(tx)
+    });
+
+    UseReactorBridgeHandle {
+        tx: tx.clone(),
+        ctr: ctr.dispatcher(),
+    }
+}
+
+/// Hook handle for the [`use_reactor_subscription`] hook.
+pub struct UseReactorSubscriptionHandle<R>
+where
+    R: 'static + Reactor,
+{
+    bridge: UseReactorBridgeHandle<R>,
+    outputs: Vec<Rc<<R::Scope as ReactorScoped>::Output>>,
+    finished: bool,
+    ctr: usize,
+}
+
+impl<R> UseReactorSubscriptionHandle<R>
+where
+    R: 'static + Reactor,
+{
+    /// Send an input to a reactor agent.
+    pub fn send(&self, msg: <R::Scope as ReactorScoped>::Input) {
+        self.bridge.send(msg);
+    }
+
+    /// Returns whether the current bridge has received a finish message.
+    pub fn finished(&self) -> bool {
+        self.finished
+    }
+
+    /// Reset the subscription.
+    ///
+    /// This disconnects the old bridge and re-connects the agent with a new bridge.
+    /// Existing outputs stored in the subscription will also be cleared.
+    pub fn reset(&self) {
+        self.bridge.reset();
+    }
+}
+
+impl<R> Clone for UseReactorSubscriptionHandle<R>
+where
+    R: 'static + Reactor,
+{
+    fn clone(&self) -> Self {
+        Self {
+            bridge: self.bridge.clone(),
+            outputs: self.outputs.clone(),
+            ctr: self.ctr,
+            finished: self.finished,
+        }
+    }
+}
+
+impl<R> fmt::Debug for UseReactorSubscriptionHandle<R>
+where
+    R: 'static + Reactor,
+    <R::Scope as ReactorScoped>::Input: fmt::Debug,
+    <R::Scope as ReactorScoped>::Output: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("UseReactorSubscriptionHandle<_>")
+            .field("bridge", &self.bridge)
+            .field("outputs", &self.outputs)
+            .finish()
+    }
+}
+
+impl<R> Deref for UseReactorSubscriptionHandle<R>
+where
+    R: 'static + Reactor,
+{
+    type Target = [Rc<<R::Scope as ReactorScoped>::Output>];
+
+    fn deref(&self) -> &Self::Target {
+        &self.outputs
+    }
+}
+
+impl<R> PartialEq for UseReactorSubscriptionHandle<R>
+where
+    R: 'static + Reactor,
+{
+    fn eq(&self, rhs: &Self) -> bool {
+        self.bridge == rhs.bridge && self.ctr == rhs.ctr
+    }
+}
+
+/// A hook to subscribe to the outputs of a [Reactor] agent.
+///
+/// All outputs sent to current bridge will be collected into a slice.
```

Comment:
Why collect them into a slice instead of providing a stream?

Reply:
There are 2 hooks for reactor and worker agents - `use_*_bridge` and `use_*_subscription`.

The `use_*_bridge` hook uses a callback-like scheme where each output from agents will call the callback function.
The `use_*_subscription` hook mimics a common [`useSubscription`](https://www.apollographql.com/docs/react/data/subscriptions/) -style hook from common React GraphQL clients.
For Subscriptions, past messages are collected into a slice. So one can do things like:


```rust
let messages_html = messages_data
    .iter()
    .map(|m| html! {
        <div key={m.id}>{m.content}</div>
    })
    .collect::<Vec<_>>();
```

Another reason to avoid a stream is to avoid a `spawn_local`.
If a stream is provided, then a future must be spawned to actively poll the stream.

If users do not need things to be collected into a slice, then they can use the `_bridge` variant with a callback.

Reply:
I see, that makes sense. It would be good to have an example in the doc comment.